Hakawai/HKWMentionsPluginV2.m

```diff
@@ -0,0 +1,1372 @@
+        }
+        return _creationStateMachine;
+    }
+}
+
+- (NSString *)pluginName {
+    return @"Mentions Creation";
+}
+
+- (BOOL)viewportLocksToTopUponMentionCreation {
+    return (self.chooserPositionMode == HKWMentionsChooserPositionModeEnclosedTop
+            || self.chooserPositionMode == HKWMentionsChooserPositionModeCustomLockTopArrowPointingUp
+            || self.chooserPositionMode == HKWMentionsChooserPositionModeCustomLockTopArrowPointingDown
+            || self.chooserPositionMode == HKWMentionsChooserPositionModeCustomLockTopNoArrow);
+}
+
+- (BOOL)viewportLocksToBottomUponMentionCreation {
+    return (self.chooserPositionMode == HKWMentionsChooserPositionModeEnclosedBottom
+            || self.chooserPositionMode == HKWMentionsChooserPositionModeCustomLockBottomArrowPointingUp
+            || self.chooserPositionMode == HKWMentionsChooserPositionModeCustomLockBottomArrowPointingDown
+            || self.chooserPositionMode == HKWMentionsChooserPositionModeCustomLockBottomNoArrow);
+}
+
+- (BOOL)viewportLocksUponMentionCreation {
+    return self.viewportLocksToTopUponMentionCreation || self.viewportLocksToBottomUponMentionCreation;
+}
+
+// Pass-through properties
+
+- (UIColor *)chooserViewBackgroundColor {
+    return self.creationStateMachine.chooserViewBackgroundColor;
+}
+
+- (void)setChooserViewBackgroundColor:(UIColor *)chooserViewBackgroundColor {
+    self.creationStateMachine.chooserViewBackgroundColor = chooserViewBackgroundColor;
+}
+
+- (UIEdgeInsets)chooserViewEdgeInsets {
+    return self.creationStateMachine.chooserViewEdgeInsets;
+}
+
+- (void)setChooserViewEdgeInsets:(UIEdgeInsets)chooserViewEdgeInsets {
+    self.creationStateMachine.chooserViewEdgeInsets = chooserViewEdgeInsets;
+}
+
+- (Class<HKWChooserViewProtocol>)chooserViewClass {
+    return self.creationStateMachine.chooserViewClass;
+}
+
+- (void)setChooserViewClass:(Class<HKWChooserViewProtocol>)chooserViewClass {
+    self.creationStateMachine.chooserViewClass = chooserViewClass;
+}
+
+/*!
+ Shows chooser view.
+ Needs to be public for integration between Hakawai and HotPot.
+ */
+- (void)showChooserView {
+    [self.creationStateMachine showChooserView];
+}
+
+/*!
+ Handles the selection from the user. This is only needed for consumers who use custom chooser view.
+ */
+- (void)handleSelectionForEntity:(id<HKWMentionsEntityProtocol>)entity {
+    [self.creationStateMachine handleSelectionForEntity:entity
+                                              indexPath:[NSIndexPath indexPathForItem:0 inSection:0]];
+}
+
+/*!
+ Gets the input made by the user for both @ mentions or # hashtag.
+ Needs to be public for integration between Hakawai and HotPot.
+ */
+- (unichar)getExplicitSearchControlCharacter {
+    return self.creationStateMachine.explicitSearchControlCharacter;
+}
+
+// TODO: Remove this call
+// JIRA: POST-14031
+- (void)textViewDidProgrammaticallyUpdate:(UITextView * _Null_unspecified)textView {
+    return;
+}
```

Comment:
does it not cause any regression?

Reply:
no. cursor movement should handle all this